### PR TITLE
asset_tracker_v2: Add documentation for the utility module

### PR DESCRIPTION
**Pull request by simensrostad**
Add documentation for the utility module.

Resolves CIA-505

### CODE REVIEW

**Comment by simensrostad** [comment]
This PR contains code changes but is labeled `ci-disable` and `DNM` for the time being. Ill remove the labels and re-run once the doc changes have been approved.

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+Watchdog
+========
+
+The module implements a watchdog library that monitors the system workqueue using the :ref:`Zephyr Watchdog driver <watchdog_api>`.
+To configure the watchdog timeout that is used, set the :ref:`CONFIG_WATCHDOG_APPLICATION_TIMEOUT_SEC <CONFIG_WATCHDOG_APPLICATION_TIMEOUT_SEC>` Kconfig option.
+The header file of the library is located at :file:`asset_tracker_v2/src/watchdog/watchdog_app.h`.
+
+If the watchdog is not fed within the :ref:`CONFIG_WATCHDOG_APPLICATION_TIMEOUT_SEC <CONFIG_WATCHDOG_APPLICATION_TIMEOUT_SEC>` timeout, a watchdog timeout occurs, causing a reboot that is initiated by the watchdog peripheral hardware unit on the nRF9160 DK.
+The watchdog library is set up to feed the :ref:`Zephyr Watchdog driver <watchdog_api>` with the system workqueue constantly at a time interval that equals half of the value specified by :ref:`CONFIG_WATCHDOG_APPLICATION_TIMEOUT_SEC <CONFIG_WATCHDOG_APPLICATION_TIMEOUT_SEC>`.
+Meaning that if the watchdog timeout is set to 60 seconds, the system workqueue feeds the watchdog every 30 seconds.
```
Comment (umapraseeda) [inline applications/asset_tracker_v2/doc/util_module.rst:42]:
```suggestion
This means that if the watchdog timeout is set to 60 seconds, the system workqueue feeds the watchdog every 30 seconds.
```

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+A reboot caused by a watchdog timeout occurs if the system workqueue is blocked and unable to feed the watchdog.
+
+Configuration options
+*********************
+
```
Comment (umapraseeda) [inline applications/asset_tracker_v2/doc/util_module.rst:47]:
```suggestion

You can set the following options to configure the utility module:

```

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+.. _CONFIG_UTIL_REBOOT_TIMEOUT:
+
+CONFIG_UTIL_REBOOT_TIMEOUT - Utility module reboot timeout
+   This option specifies the timeout after an irrecoverable error has been reported to the utility module, that the module initiates a reboot.
```
Comment (umapraseeda) [inline applications/asset_tracker_v2/doc/util_module.rst:51]:
```suggestion
   This option specifies the timeout within which the utility module initiates a reboot, after an irrecoverable error has been reported to the module.
```

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+* :ref:`event_manager`
+* :file:`asset_tracker_v2/src/watchdog/watchdog_app.h`
+
+API documentation
+*****************
+
+| Header file: :file:`asset_tracker_v2/src/events/util_module_event.h`
+| Source files: :file:`asset_tracker_v2/src/events/util_module_event.c`
+                :file:`asset_tracker_v2/src/modules/util_module.c`
```
Comment (umapraseeda) [inline applications/asset_tracker_v2/doc/util_module.rst:87]:
```suggestion
| Source files: :file:`asset_tracker_v2/src/events/util_module_event.c`,               :file:`asset_tracker_v2/src/modules/util_module.c`
```

**Review by jtguggedal — APPROVED** [review]

**Review by umapraseeda — APPROVED** [review]

File: applications/asset_tracker_v2/src/modules/Kconfig.util_module [hunk 5]
```diff
@@ -10,13 +10,14 @@ menuconfig UTIL_MODULE
 
 if UTIL_MODULE
 
-config REBOOT_TIMEOUT
```
Comment (tejlmand) [inline applications/asset_tracker_v2/src/modules/Kconfig.util_module:13]: Is a symbol rename really needed ?
And if it is, could the old name be deprecated first to give users time to adjust.

Comment (simensrostad) [reply applications/asset_tracker_v2/src/modules/Kconfig.util_module:13]: No, it's not strictly needed. I removed the name change for now.

File: applications/asset_tracker_v2/doc/modules_common.rst [hunk 6]
```diff
@@ -0,0 +1,26 @@
+.. _api_modules_common:
+
+Modules common API
+##################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The Modules common API provides functionality to administer mechanisms in the application architecture, and consist of functions that are shared between modules.
```
Comment (divipillai) [inline applications/asset_tracker_v2/doc/modules_common.rst:10]:
```suggestion
The Modules common API provides the functionality to administer mechanisms in the application architecture, and consist of functions that are shared between modules.
```

File: applications/asset_tracker_v2/doc/modules_common.rst [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+The API supports the following:
```
Comment (divipillai) [inline applications/asset_tracker_v2/doc/modules_common.rst:11]:
```suggestion
The API supports the following functionalities:
```

File: applications/asset_tracker_v2/doc/modules_common.rst [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+* Registering and starting a module using :c:func:`modules_start`.
```
Comment (divipillai) [inline applications/asset_tracker_v2/doc/modules_common.rst:13]:
```suggestion
* Registering and starting a module using :c:func:`module_start`.
```

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+This section documents the various features implemented by the module.
+
+Application Shutdown
+====================
+
+When an irrecoverable error or a successful FOTA update is reported to the utility module, it initiates a shutdown sequence and ensures a graceful shutdown of the application.
+
+During initialization, each module registers if it supports graceful shutdown.
+This is done by calling the :c:func:`module_start` function in the :ref:`Modules common API <api_modules_common>` and passing a configuration value in the :c:struct:`module_data` structure for each module that registers for graceful shutdown.
```
Comment (divipillai) [inline applications/asset_tracker_v2/doc/util_module.rst:23]:
```suggestion
This is done by calling the :c:func:`module_start` function in the :ref:`api_modules_common` and passing a configuration value in the :c:struct:`module_data` structure for each module that registers for graceful shutdown.
```

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+This then expands an internal list in the :ref:`Modules common API <api_modules_common>`.
```
Comment (divipillai) [inline applications/asset_tracker_v2/doc/util_module.rst:24]:
```suggestion
This then expands an internal list in the :ref:`api_modules_common`.
```

File: applications/asset_tracker_v2/doc/util_module.rst [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+#. The utility module receives the error event and starts a reboot timer that times out after the duration specified by the :ref:`CONFIG_REBOOT_TIMEOUT <CONFIG_REBOOT_TIMEOUT>` option. When this timer expires, the application reboots in any case.
+#. The utility module sends out a shutdown request, :c:enum:`UTIL_EVT_SHUTDOWN_REQUEST`, that is handled individually by each module in the system that supports shutdown.
+#. When each module has handled the shutdown request, it acknowledges the shutdown by sending an event suffixed with ``EVT_SHUTDOWN_READY``. For example, :c:enum:`CLOUD_EVT_SHUTDOWN_READY`.
+#. When the utility module receives a shutdown acknowledgment, for each incoming shutdown acknowledgment, the aforementioned list in the :ref:`Modules common API <api_modules_common>` is checked using :c:func:`modules_shutdown_register`. When all the modules in the list have acknowledged the request, the utility module reboots the application at a much earlier point than the time set initially by the reboot timer.
```
Comment (divipillai) [inline applications/asset_tracker_v2/doc/util_module.rst:32]:
```suggestion
#. When the utility module receives a shutdown acknowledgment, for each incoming shutdown acknowledgment, the aforementioned list in the :ref:`api_modules_common` is checked using :c:func:`modules_shutdown_register`. When all the modules in the list have acknowledged the request, the utility module reboots the application at a much earlier point than the time set initially by the reboot timer.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 12]
```diff
@@ -66,11 +66,8 @@ nRF9160: Asset Tracker v2
 * Added support for atmospheric pressure readings retrieved from the BME680 sensor on Thingy:91.
 * Fixed an issue where PSM could be requested from the network even though it was disabled in Kconfig.
 * Added new documentation for Asset Tracker v2 :ref:`asset_tracker_v2_modem_module`.
-
-* Added:
-
-  * Support for A-GPS filtered ephemerides.
-
+* Added support for A-GPS filtered ephemerides.
+* Added new documentation for :ref:`asset_tracker_v2_util_module`.
```
Comment (divipillai) [inline doc/nrf/releases/release-notes-changelog.rst:70]:
```suggestion
* Added new documentation for :ref:`asset_tracker_v2_util_module` and :ref:`api_modules_common`.
```

**Comment by simensrostad** [comment]
@divipillai Review comments addressed, thanks.

**Review by divipillai — APPROVED** [review]

**Pull request merged**